domains/primitives/messenger-host-functions/src/host_functions.rs

```diff
@@ -0,0 +1,139 @@
+
+/// Trait to query messenger specific details.
+pub trait MessengerHostFunctions: Send + Sync {
+    /// Returns the storage key for the given request.
+    fn get_storage_key(&self, req: StorageKeyRequest) -> Option<Vec<u8>>;
+}
+
+sp_externalities::decl_extension! {
+    pub struct MessengerExtension(Arc<dyn MessengerHostFunctions>);
+}
+
+impl MessengerExtension {
+    /// Create a new instance of [`MessengerExtension`].
+    pub fn new(inner: Arc<dyn MessengerHostFunctions>) -> Self {
+        Self(inner)
+    }
+}
+
+/// Implementation of Messenger host function.
+pub struct MessengerHostFunctionsImpl<Block, Client, DomainBlock, Executor> {
+    consensus_client: Arc<Client>,
+    executor: Arc<Executor>,
+    _phantom: PhantomData<(Block, DomainBlock)>,
+}
+
+impl<Block, Client, DomainBlock, Executor>
+    MessengerHostFunctionsImpl<Block, Client, DomainBlock, Executor>
+{
+    pub fn new(consensus_client: Arc<Client>, executor: Arc<Executor>) -> Self {
+        MessengerHostFunctionsImpl {
+            consensus_client,
+            executor,
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<Block, Client, DomainBlock, Executor>
+    MessengerHostFunctionsImpl<Block, Client, DomainBlock, Executor>
+where
+    Block: BlockT,
+    DomainBlock: BlockT,
+    Client: HeaderBackend<Block> + ProvideRuntimeApi<Block>,
+    Client::Api: DomainsApi<Block, DomainBlock::Header>,
+{
+    fn get_domain_runtime_code(
+        &self,
+        consensus_block_hash: Block::Hash,
+        domain_id: DomainId,
+    ) -> Option<Vec<u8>> {
+        let runtime_api = self.consensus_client.runtime_api();
+        // Use the parent hash to get the actual used domain runtime code
+        // TODO: update once we can get the actual used domain runtime code by `consensus_block_hash`
+        let consensus_block_header = self
+            .consensus_client
+            .header(consensus_block_hash)
+            .ok()
+            .flatten()?;
+        runtime_api
+            .domain_runtime_code(*consensus_block_header.parent_hash(), domain_id)
+            .ok()
+            .flatten()
+    }
+}
+
+impl<Block, Client, DomainBlock, Executor> MessengerHostFunctions
+    for MessengerHostFunctionsImpl<Block, Client, DomainBlock, Executor>
+where
+    Block: BlockT,
+    Block::Hash: From<H256>,
+    DomainBlock: BlockT,
+    Client: HeaderBackend<Block> + ProvideRuntimeApi<Block>,
+    Client::Api: MessengerApi<Block, NumberFor<Block>> + DomainsApi<Block, DomainBlock::Header>,
+    Executor: CodeExecutor + RuntimeVersionOf,
+{
+    fn get_storage_key(&self, req: StorageKeyRequest) -> Option<Vec<u8>> {
+        let best_hash = self.consensus_client.info().best_hash;
+        let runtime_api = self.consensus_client.runtime_api();
+        match req {
+            StorageKeyRequest::ConfirmedDomainBlockStorageKey(domain_id) => runtime_api
+                .confirmed_domain_block_storage_key(best_hash, domain_id)
+                .map(Some),
```

Comment:
I don't think I understand this host function either, comment is the same as for previous hash functions

Reply:
I see the benefits of decoupling, but I'm confused about what this has to do with domains, this seems to call consensus client runtime API :thinking:

Reply:
So the `domain_proof` is a storage proof derived from `LatestConfirmedDomainBlock` on consensus runtime. We are using consensus runtime api to get that storage key while verifying the `domain_proof` and the use src domain runtime to fetch the `Outbox` and `InboxResponse` storage keys to complete the XDM message verification